[FLINK-12607][rest] Introduce a REST API that returns the maxParallelism

bytesandwich opened the pull request:
## What is the purpose of the change

Add max parallelism for reporting purposes to the REST api.

## Brief change log

Add max-parallelism to:
- JobVertexDetailsHandler puts maxaParallelism into JobVertexDetailsInfo from jobVertex
- JobDetailsHandler pus:
    - maxParallelism into JobDetailsInfo from ExecutionGraph's archived execution config
    - maxParallelism into JobVertexDetailsInfo from AccessExecutionJobVertex

- ArchivedExecutionConfig

- JobDetailsInfo as "job-max-parallelism"
    - JobDetailsInfo.JobVertexDetailsInfo as "max-parallelism"
- JobVertexDetailsInfo as "max-parallelism"

All have max parallelism added to equals and hash (except ArchivedExecutionConfig)

Resets ArchivedExecutionConfig's `serialVersionUID`

also adds to angular type definitions.



*(For example: This pull request makes task deployment go through the blob server, rather than through RPC. That way we avoid re-transferring them on each deployment (during recovery).)*


## Verifying this change

This change is already covered by existing tests, such as `JobDetailsInfoTest` and `JobVertexDetailsInfoTest`. As I'm developing on windows linux subsystem, my maven build works but I am unable to run tests locally. Is there an automation that runs them to check this PR? 

I did manually run the REST api against a local cluster from the bin/start-cluster.sh  of TopSpeedWindowing example with `env.getConfig().setMaxParallelism(4)` and `.setMaxParallelism(2);` on an operator

In both `/jobs`and `/jobs/:id` I saw that "job-max-parallelism" was set to 4 and "max-parallelism" was set to "2" for the specific vertex and "4" for others.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no but I feel like it should?
  - The serializers: yes the ArchivedExecutionConfig's java serializability
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? slightly
  - If yes, how is the feature documented? Needs a follow up for the REST api docs.

zentol commented on .gitignore:
```diff
@@ -12,6 +12,7 @@ target
 tmp
 *.class
 *.iml
+*.run
```
What is this ignore for?

bytesandwich replied:
That is for an intellij run profile.

zentol replied:
let's move that into a separate hotfix commit

bytesandwich replied:
Done.

zentol commented on flink-runtime-web/web-dashboard/src/app/interfaces/job-detail.ts:
```diff
@@ -50,6 +50,8 @@ export interface JobDetailInterface {
   'end-time': number;
   duration: number;
+  parallelism: number;
+  'job-max-parallelism': number;
```
it should be fine to call this `max-parallelism`

bytesandwich replied:
Done

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/JobVertexDetailsInfo.java:
```diff
@@ -41,6 +41,7 @@
     public static final String FIELD_NAME_VERTEX_ID = "id";
     public static final String FIELD_NAME_VERTEX_NAME = "name";
     public static final String FIELD_NAME_PARALLELISM = "parallelism";
+    public static final String FIELD_NAME_MAX_PARALLELISM = "max-parallelism";
```
new fields should follow `camelCase` (dashes are a legacy format that we're trying to phase out)

bytesandwich replied:
Done

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/JobDetailsInfo.java:
```diff
@@ -59,6 +59,8 @@
 
     public static final String FIELD_NAME_DURATION = "duration";
 
+    public static final String FIELD_NAME_MAX_PARALLELISM = "job-max-parallelism";
```
same as in other files; remove "job" prefix and adjust case

bytesandwich replied:
Done

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/messages/JobVertexDetailsInfoTest.java:
```diff
@@ -83,10 +83,12 @@ protected JobVertexDetailsInfo getTestResponseInstance() throws Exception {
                 "jobVertex" + random.nextLong(),
-                random.nextInt(),
+                parallelism,
+                2 * parallelism,
```
This could overflow leading to a negative maxParallelism which may break an assertion somewhere at some point

bytesandwich replied:
Good point!

bytesandwich replied:
Done

zentol reviewed (changes requested):
This looks good overall but I had some smaller comments.

While the REST API documentation is indeed generated it must be triggered manually. You can find instructions for doing so [here](https://github.com/apache/flink/blob/master/flink-docs/README.md).
You also need to regenerate the REST API snapshot that is used for compatibility checks:
```
[ERROR]   RestAPIStabilityTest.testDispatcherRestAPIStability:94->assertCompatible:156 The API was modified in a compatible way, but the snapshot was not updated. To update the snapshot, re-run this test with -Dgenerate-rest-snapshot being set.
```

bytesandwich commented:
I'm still working on getting those doc generation and checkpoint commands going.

zentol commented on flink-runtime-web/web-dashboard/src/app/interfaces/job-detail.ts:
```diff
@@ -50,6 +50,8 @@ export interface JobDetailInterface {
   'start-time': number;
   'end-time': number;
   duration: number;
+  parallelism: number;
```
hmm the `JobDetailsInfo` doesn't seem to contain a `parallelism` field?

bytesandwich replied:
Ah, true. Done.

zentol commented:
The generation can be a bit finicky; essentially you need to run these commands locally and commit the generated files.

```
mvn package -pl flink-runtime,flink-docs -Dfast -Dgenerate-rest-docs
mvn test -pl flink-runtime-web -Pskip-webui-build -Dfast -Dtest=RestAPIStabilityTest -Dgenerate-rest-snapshot
```
If you run into issues with these please ping me.

bytesandwich commented:
Thank you for your quick and attentive code review, @zentol! I also appreciate that you sent those commands. 

I'm having a bit of trouble running tests and node builds under windows WSL linux. So, I ran
```
# From the link
mvn package -Dgenerate-rest-docs -pl flink-docs -am -nsu -DskipTests -Pskip-webui-build
```

And I wasn't able to get `mvn package` without `-DskipTests` or `mvn test` to work, but I was able to run that specific test in intellij for some reason.

I think I cleaned it up, rebased, and ran pretty much those commands? PTAL. I think I addressed all the comments you've made so far.

zentol commented on flink-runtime-web/src/test/resources/rest_api_v1.snapshot:
```diff
@@ -1,3195 +1,3204 @@
-{
-  "calls" : [ {
-    "url" : "/cluster",
```
Seems like we got a line-ending problem here.

Please cherry-pick https://github.com/zentol/flink/commit/207cdf2cadbbbe9e9158164c81d492d924b531e1, and generate the snapshot again (or convert the line-endings manually, whichever is easier for you)

bytesandwich replied:
Thanks! Done.

bytesandwich commented:
My first commit to Flink! Thank you for your review @zentol 😃 !
